vdev/src/features.rs

```diff
@@ -0,0 +1,128 @@
+use std::{collections::BTreeSet, collections::HashMap, ffi::OsStr, fs, path::Path};
+
+use anyhow::{bail, Context, Result};
+use serde::Deserialize;
+use serde_json::Value;
+
+type ComponentMap = HashMap<String, Component>;
+
+// Use a BTree to keep the results in sorted order
+type FeatureSet = BTreeSet<String>;
+
+/// This is a ersatz copy of the Vector config, containing just the elements we are interested in
+/// examining. Everything else is thrown away.
+#[derive(Deserialize)]
+pub struct VectorConfig {
+    api: Option<Value>,
+    enterprise: Option<Value>,
+
+    #[serde(default)]
+    sources: ComponentMap,
+    #[serde(default)]
+    transforms: ComponentMap,
+    #[serde(default)]
+    sinks: ComponentMap,
+}
+
+#[derive(Deserialize)]
+struct Component {
+    r#type: String,
+}
+
+macro_rules! mapping {
+    ( $( $key:ident => $value:ident, )* ) => {
+        HashMap::from([
+            $( (stringify!($key), stringify!($value)), )*
+        ])
+    };
+}
+
+pub fn load_and_extract(filename: impl AsRef<Path>) -> Result<String> {
+    let filename = filename.as_ref();
+    let config =
+        fs::read_to_string(filename).with_context(|| format!("failed to read {filename:?}"))?;
+
+    let config: VectorConfig = match filename.extension().and_then(OsStr::to_str) {
+        None => bail!("Invalid filename {filename:?}, no extension"),
+        Some("json") => serde_json::from_str(&config)?,
+        Some("toml") => toml::from_str(&config)?,
+        Some("yaml") => serde_yaml::from_str(&config)?,
+        Some(_) => bail!("Invalid filename {filename:?}, unknown extension"),
+    };
+
+    Ok(from_config(config))
+}
+
+pub fn from_config(config: VectorConfig) -> String {
+    // Mapping of component names to feature name exceptions.
```

Comment:
Some of these seem like we could actually rename to avoid this manual mapping, have we considered/planned to do so, or am I missing a reason why we couldn't for renames like `sampler => sample`.

Reply:
This particular one is no longer needed, as the `sampler` alias no longer exists. Same for `swimlanes`. I'll check the others. Some will have to remain, like the group features for `prometheus` and `gcp`.